services/scan/src/store.ts

```diff
@@ -322,100 +394,72 @@ export class Store {
+    }
+
+    return precinctSelectionParseResult.ok();
   }
 
   /**
    * Sets the current precinct `scan` is accepting ballots for. Set to
    * `undefined` to accept from all precincts (this is the default).
    */
   setPrecinctSelection(precinctSelection?: PrecinctSelectionType): void {
-    this.setConfig(ConfigKey.PrecinctSelection, precinctSelection);
+    if (this.hasElection()) {
+      if (!precinctSelection) {
+        this.client.run('update election set precinct_selection = null');
+      } else {
+        this.client.run(
+          'update election set precinct_selection = ?',
+          JSON.stringify(precinctSelection)
+        );
+      }
+    } else {
+      throw new Error('Cannot set precinct selection without an election.');
+    }
   }
 
   /**
    * Gets the current polls state (open, paused, closed initial, or closed final)
    */
   getPollsState(): PollsStateType {
-    return this.getConfig(
-      ConfigKey.PollsState,
-      'polls_closed_initial',
-      PollsStateSchema
-    );
-  }
-
-  /**
-   * Sets the current polls state
-   */
-  setPollsState(pollsState: PollsStateType): void {
-    this.setConfig(ConfigKey.PollsState, pollsState);
-  }
-
-  /**
-   * Gets a config value by key.
-   */
-  private getConfig<T>(key: ConfigKey, schema: z.ZodSchema<T>): T | undefined;
-  private getConfig<T>(
-    key: ConfigKey,
-    defaultValue: T,
-    schema: z.ZodSchema<T>
-  ): T;
-  private getConfig<T>(
-    key: ConfigKey,
-    defaultValueOrSchema: z.ZodSchema<T> | T,
-    maybeSchema?: z.ZodSchema<T>
-  ): T | undefined {
-    debug('get config %s', key);
-
-    const row = this.client.one<[string]>(
-      'select value from configs where key = ?',
-      key
-    ) as { value: string } | undefined;
-
-    let defaultValue: T | undefined;
-    let schema: z.ZodSchema<T>;
-
-    if (maybeSchema) {
-      defaultValue = defaultValueOrSchema as T;
-      schema = maybeSchema;
-    } else {
-      schema = defaultValueOrSchema as z.ZodSchema<T>;
-    }
-
-    if (typeof row === 'undefined') {
-      debug('returning default value for config %s: %o', key, defaultValue);
-      return defaultValue;
-    }
-
-    const result = safeParseJson(row.value, schema);
+    const electionRow = this.client.one('select polls_state from election') as
+      | { polls_state: PollsStateType }
+      | undefined;
 
-    if (result.isErr()) {
-      debug('failed to validate stored config %s: %s', key, result.err());
-      return undefined;
+    if (!electionRow) {
+      // we will not skip the check by default once an election is defined
+      return 'polls_closed_initial';
     }
 
-    const value = result.ok();
-    let inspectedResult = inspect(value, false, 2, true);
-    if (inspectedResult.length > 200) {
-      inspectedResult = `${inspectedResult.slice(0, 199)}…`;
-    }
-    debug('returning stored value for config %s: %s', key, inspectedResult);
-    return value;
+    return electionRow.polls_state;
```

Comment:
Validate it at all?

Reply:
It seems unnecessary in that typescript is validating the value entering the DB, and it only has one entry point, so it should always be correct. Is that myopic?

Also, if we want to validate it, why not validate it at the DB level with a `CHECK` in the schema?

Reply:
Data in a database can get out of sync with the code that uses it. For example, say we had this:

```ts
interface Person {
  firstName: string;
  lastName: string
}
```

And we stored such a value in the database as a JSON blob. Then, sometime later, we changed the code like so:

```ts
interface Person {
  givenName: string;
  familyName: string;
}
```

We could do `const person: Person = JSON.parse(row);` and say that the data should be correct because when we inserted the data it was a TypeScript-validated `Person` and now we're loading it as a `Person` so it's fine, right? But since the `Person` type changed we can no longer make any guarantees. That's why, in general, when we load data from external sources I like to validate it even if we "own" the data and put it there ourselves.

Depending on your desired behavior you can throw an exception, ignore the invalid data, migrate it to the new format, or something else. But if you never check you'll never know it's wrong until something blows up or misbehaves later.

Reply:
In this particular case you called out, I ended up adding the validation. Your explanation makes sense to me!